src/device/array.jl

```diff
@@ -0,0 +1,155 @@
+# Contiguous on-device arrays
+
+export VeDeviceArray, VeDeviceVector, VeDeviceMatrix, ldg
+
+
+## construction
+
+"""
+    VeDeviceArray(dims, ptr)
+    VeDeviceArray{T}(dims, ptr)
+    VeDeviceArray{T,N}(dims, ptr)
+    VeDeviceArray{T,N,A}(dims, ptr)
+
+Construct an `N`-dimensional dense VEDA device array with element type `T` wrapping a
+pointer, where `N` is determined from the length of `dims` and `T` is determined from the
+type of `ptr`. `dims` may be a single scalar, or a tuple of integers corresponding to the
+lengths in each dimension). If the rank `N` is supplied explicitly as in `Array{T,N}(dims)`,
+then it must match the length of `dims`. The same applies to the element type `T`, which
+should match the type of the pointer `ptr`.
+"""
+VeDeviceArray
+
+# NOTE: we can't support the typical `tuple or series of integer` style construction,
+#       because we're currently requiring a trailing pointer argument.
+
+struct VeDeviceArray{T,N,A} <: AbstractArray{T,N}
+    shape::Dims{N}
+    ptr::LLVMPtr{T,A}
```

Comment:
IIUC this should be a `API.VEDAdeviceptr` and we need to convert it device side to a normal `Ptr`. 

To quote @efocht 

>if you get to memory pointers, there is a peculiarity. You can deal with VEDAdeviceptr that is allocated by VEDA calls, eg. vedaMemAllocAsync(&ptr, size), which returns a "virtual" address immediately, although the allocation will be deffered. That type of pointer needs to be dereferenced in the VE kernel by vedaMemPtr() which makes a real pointer out of it (allocation should be finished at that point).
OTOH, if you allocate directly, with malloc in kernels on VE side, those addresses are real pointers.

Reply:
That shouldn't matter, though, because we just need to do a conversion on the host side from `VEDAdeviceptr` to `LLVMPtr` (and choose the right addrspace based on the type of allocation done). Does that sound correct? It was my understanding that we use an `LLVMPtr` to ensure we pass the correct addrspaces to LLVM.

Reply:
IIUC we don't need LLVMPtr on the VE. There are no addresspaces to keep track of, and we need to devirtualize the pointer on the device.